### PR TITLE
[java] Update rule UnnecessaryImport

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Part of #2701 

## Related issues

- Fixes #

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/3009496d2425496fa7325aa2276b5031/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 48833 violations,<br />
introduces 26251 new violations, 5 new errors and 0 new configuration errors,<br />
removes 151342 violations, 27 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/3009496d2425496fa7325aa2276b5031/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 1]
```diff
@@ -157,7 +158,8 @@ public class Foo {
         ]]></code>
     </test-code>
 
-    <test-code>
+    <test-code regressionTest="false">
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:161]: I disabled this test, as we now need a classpath to correctly identify imported static methods/fields. This should make the rule more precise when the classpath is full, but maybe more prone to FPs when the classpath is improperly configured.

Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:161]: I think, that's fine. At some point we need to require that PMD is called properly...

**Comment by adangel** [comment]
Thanks for updating this rule - it looks good. While looking through the regression tester report, I found some possible false positives:

a) https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeTestStaticImports.java#L20
I think, this is explainable by incomplete classpath: the referenced class "InputIllegalType" is in the same package, but all this source is under "src/test/resources" - so it is not compiled. Maybe it's worth to specially try to compile these additional sources from checkstyle - should be doable in https://github.com/pmd/pmd/blob/19529542dbdef7815825cfcdbd6461f5e8345a82/.ci/files/project-list.xml#L27 by adding some custom "javac" calls... but that's for another PR.

b) https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/java/com/puppycrawl/tools/checkstyle/AbstractXmlTestSupport.java#L22
com.google.common.truth.Truth.assertThat is claimed to be unnecessary. But I think it is indeed used at [Line 230](https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/java/com/puppycrawl/tools/checkstyle/AbstractXmlTestSupport.java#L230).
Maybe we have a problem with statically imported methods?
In spring framework, there is a similar case with [org.assertj.core.api.Assertions.within](https://github.com/spring-projects/spring-framework/blob/v5.3.13/spring-beans/src/test/java/org/springframework/beans/propertyeditors/CustomEditorTests.java#L55).

Unless we have a problem with our classpath, since both methods would come in through a 3rd party library...

c) https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/solaris/classes/sun/nio/ch/DevPollSelectorImpl.java#L40
Unused import "sun.nio.ch.DevPollArrayWrapper.NUM_POLLFDS". 
That's probably due to the fact, that the class DevPollArrayWrapper is not on the classpath. It's from the package "sun.nio", so it's a platform specific implementation... in this case for solaris...

d) https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/java/util/concurrent/SubmissionPublisher.java#L49
Unused import "java.util.concurrent.Flow.Subscription"
In this case, it should be on the classpath, shouldn't it? I'm not sure at the moment, how we deal with jdk classes - all the types are loaded via AsmSymbolResolver, right? Not sure, if we also have the jdk itself on the auxclasspath....

In this (big) file SubmissionPublisher, the interface "Subscription" is mostly referenced via "Flow.Subscription", but on [Line 886](https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/java/util/concurrent/SubmissionPublisher.java#L886) just with "Subscription".

**Comment by oowekyala** [comment]
Thanks for the review! I think I'll add additional logic to be more conservative when an import is unresolved. The rule already handles unresolved type imports gracefully, thanks to the unresolved symbols of the symbol API. I think we should also be more intelligent with unresolved static imports.

**Comment by oowekyala** [comment]
Actually let's do this in a followup PR. I'll investigate the FP (d) you found in this PR.

**Comment by oowekyala** [comment]
Note: new FP https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/macosx/classes/sun/nio/ch/KQueuePort.java#L36

**Comment by oowekyala** [comment]
I think the new FP should be fixed with #3748. This PR is ready to merge IMO

**Pull request merged**